### PR TITLE
Serving multiple environments

### PR DESCRIPTION
**Pull request by Luke100000**
`RemoteEnvironmentService` is now a server, providing a separate env for each connection, closing them when connection is lost or requested.

Each env runs in a subprocess, thus, the server scales well and can, if requested, run different branches/commits.

### CODE REVIEW

File: exploration/start_remote_environment.py
Comment (Luke100000) [inline exploration/start_remote_environment.py:65]: I'm not 100% happy with merging args and settings here but splitting them up just adds complexity with minimal benefit either.

File: src/remote_gym/remote_environment.py
Comment (Luke100000) [inline src/remote_gym/remote_environment.py]: Not happy with omegaconf here either, Remote_args is only a Omega dict because its directly passed from the environment space. Thus, the whole omegaconf is a "hidden" dependency caused by another repo.

File: src/remote_gym/remote_environment.py
Comment (Luke100000) [inline src/remote_gym/remote_environment.py]: I dislike this, but wasn't able to pull that into where the omegaconf is passed in the first place. Imo, we should be able to safely assume a dict here.

Comment (alexander-zap) [reply src/remote_gym/remote_environment.py]: If the typing here is really DictConfig and not a dict, this should be reflected in the typing at the very top in the class instantiation arguments.

Currently it's `remote_args: dict = {}`.

File: exploration/test_repo_manager.py
Comment (alexander-zap) [inline exploration/test_repo_manager.py]: How is this test different from `test_update_default()`?
Or is the test here, that you execute it twice?

If yes, then this test does not work *on its own* and you should add another same line.

File: src/remote_gym/repo_manager.py
Comment (alexander-zap) [inline src/remote_gym/repo_manager.py:1]: Did you write this completely yourself? 
If yes, well done 👍  
If no, then add a stackoverflow link in a comment, just so maintainers know where to get started in case something doesn't work.

File: src/remote_gym/repo_manager.py
Comment (alexander-zap) [inline src/remote_gym/repo_manager.py:26]: Is this the place where it differentiates between tags and commit hashes?

If yes, then it might make sense to add to the documentation of `RepoManager.get`, that commit hashes are also supported. Currently the documentation and parameter name `tag` do not indicate that passing a branch is possible.

Comment (alexander-zap) [reply src/remote_gym/repo_manager.py:26]: Also, it seems (based on the tests) that you support branch names and commit hashes. I have not seen where tag support is implemented.

File: exploration/test_repo_manager.py
Comment (alexander-zap) [inline exploration/test_repo_manager.py:1]: If tags are supported, also create a test for it.

File: src/remote_gym/remote_environment.py
Comment (alexander-zap) [reply src/remote_gym/remote_environment.py]: But I agree with you, we should be able so assume it's a dict. Is there no way to fix this in the place it is passed?

File: exploration/start_remote_environment.py
Comment (alexander-zap) [reply exploration/start_remote_environment.py:65]: I would say readability is a nice benefit. Splitting them would be also easy to implement, since `remote_environment_manager.create_gym_environment` has a logical split.

File: src/remote_gym/remote_environment.py
Comment (Luke100000) [reply src/remote_gym/remote_environment.py]: > If the typing here is really DictConfig

It originates from the config/config_reading, however, I believe its safe to convert before passing params. Will append that to the respective PR and remove Omegaconf here.

File: src/remote_gym/repo_manager.py
Comment (Luke100000) [reply src/remote_gym/repo_manager.py:1]: Yes, my hack :)

File: src/remote_gym/repo_manager.py
Comment (Luke100000) [reply src/remote_gym/repo_manager.py:26]: Branches, tags and commit ids are all valid references, no explicit support is required here afaik. Branches however, unlike tags for some reasons, can be local. We don't want that, thus the "origin/" test to make sure it always uses the one from the origin.

File: src/remote_gym/remote_environment.py
Comment (alexander-zap) [inline src/remote_gym/remote_environment.py]: Possible "hard-coded" fields should be documented here (e.g., `repo`).

File: src/remote_gym/remote_environment.py
Comment (alexander-zap) [reply src/remote_gym/remote_environment.py]: Yeah, `params` of EnvironmentConfig is just of type dict and not touched by OmegaConf. It should be safe to convert back.

File: src/remote_gym/remote_environment_management.py
Comment (alexander-zap) [inline src/remote_gym/remote_environment_management.py]: I think at this point we might want to rename this method to `create_environment_in_server` or something like that.

File: src/remote_gym/remote_environment_management.py
Comment (alexander-zap) [inline src/remote_gym/remote_environment_management.py]:
```suggestion
        default_args: The arguments passed to the entrypoint/method which creates the environment.
```

Comment (alexander-zap) [reply src/remote_gym/remote_environment_management.py]: Also it might make sense here to document some keys which are "must-haves" (e.g. `entrypoint` itself). Especially since some keys are not allowed to be overwritten (e.g., `repo`).

Comment (alexander-zap) [inline src/remote_gym/remote_environment_management.py]: Again, I think here it makes the most sense to split server_args and environment_args, since overwriting server_args has very different semantic from overwriting environment_args. Overwriting server_args means changing the environment creation code itself (since we pull a new version of the repo), while overwriting environment_args just changes the environment initialization through parameters.

Comment (alexander-zap) [reply src/remote_gym/remote_environment_management.py]: So basically explain possible fields, same as I commented in `remote_environment.py`.

Comment (alexander-zap) [inline src/remote_gym/remote_environment_management.py]: Also it is important to explain that `entrypoint` needs a method called "create_environment".

File: src/remote_gym/remote_environment_management.py
Comment (alexander-zap) [inline src/remote_gym/remote_environment_management.py:243]: This is unexpected behavior. We should at least log something here before closing the environment.

File: src/remote_gym/remote_environment.py
Comment (Luke100000) [reply src/remote_gym/remote_environment.py]: I moved this to a separate typed dict as it occurs on several places.

File: src/remote_gym/remote_environment_management.py
Comment (alexander-zap) [reply src/remote_gym/remote_environment_management.py:243]: Nevermind, it's for closing. Didn't see it.

Comment (Luke100000) [reply src/remote_gym/remote_environment_management.py:243]: It logs it in the manager, `logging.info(f"Destroyed environment for user {user} ({len(self.environments)} total active)")`
None is an internal signal to terminate the loop.

File: src/remote_gym/remote_environment_management.py
Comment (alexander-zap) [inline src/remote_gym/remote_environment_management.py]: This is very hacky on both sides (here in `ProcessedEnv` as well as `run_env_loop` where you handle the Exception by passing None twice and pass the error message afterwards).

Comment (alexander-zap) [reply src/remote_gym/remote_environment_management.py]: I will think about a better solution, haven't come up with anything better yet.

Comment (alexander-zap) [reply src/remote_gym/remote_environment_management.py]: It might make sense to use a third queue (or maybe something like a bidirectional Pipe even would fit better) to communicate meta information, which both processes always check. 

From Env side it would make sense to pass an initialization dict object in the beginning (with "action_spec" and "observation_spec" as keys) and then check for further objects in case the main process asks for closing. 

And the loop can use it to send back an error messages if required, which the Env is checking for regularly.

File: src/remote_gym/remote_environment_management.py
Comment (alexander-zap) [inline src/remote_gym/remote_environment_management.py]: Could be called something like `available_environment_ids` for more clarity.

File: src/remote_gym/remote_environment_management.py
Comment (Luke100000) [reply src/remote_gym/remote_environment_management.py]: Notice that you then need to handle timeouts etc when fetching the specs, which is what I wanted to avoid.

File: src/remote_gym/remote_environment_management.py
Comment (alexander-zap) [reply src/remote_gym/remote_environment_management.py:243]: Got it 👍 
Yeah, I was surprised to see the `None` but later found in `ProcessedEnv` in which cases reset turns to None.
But as said in the comment about hackyness of communication: Maybe we should use a third queue and use better messages, instead of queuing "special" values to signal certain situation between processes.

File: src/remote_gym/remote_environment_management.py
Comment (alexander-zap) [inline src/remote_gym/remote_environment_management.py]: Is this used by the `create_gym_environment` method?

Comment (alexander-zap) [reply src/remote_gym/remote_environment_management.py]: Or is the idea here that the environment creation function `create_environment` *might* want to use it?

File: src/remote_gym/remote_environment_management.py
Comment (alexander-zap) [inline src/remote_gym/remote_environment_management.py]: I think custom entrypoints do not pose an additional security issue *if you do not permit usage of custom repositories*.
In this case changing the entrypoint is the same as changing the code of the entrypoint file.

Unless it is possible to point to files outside the repository, because in that case it indeed is a security risk 😅

File: src/remote_gym/remote_environment_management.py
Comment (alexander-zap) [inline src/remote_gym/remote_environment_management.py:291]: Is there a use case right now for having an env_id attribute? It seems like most management is done with the `RemoteEnvironmentServicer.environments` data structure.

File: src/remote_gym/remote_environment_management.py
Comment (alexander-zap) [inline src/remote_gym/remote_environment_management.py]: I made a typo in the class name back when I created this class. It should be `Servicer`.

File: src/remote_gym/remote_environment_management.py
Comment (alexander-zap) [inline src/remote_gym/remote_environment_management.py]: What do you mean?

File: src/remote_gym/remote_environment_management.py
Comment (alexander-zap) [inline src/remote_gym/remote_environment_management.py]: I don't think you should do this when an agent leaves the world (just based on the protocol). In our case, an experiment ending should also send `message_type == "destroy_world"`, so no need to do it here additionally.

File: src/remote_gym/remote_environment_management.py
Comment (alexander-zap) [inline src/remote_gym/remote_environment_management.py]: Also, shouldn't this be done on `message_type == "create_world"` instead?

File: src/remote_gym/remote_environment_management.py
Comment (alexander-zap) [reply src/remote_gym/remote_environment_management.py]: Are you referencing here the fact that you add a callback and *additionally* call this on "destroy_world"?

**Review by alexander-zap — CHANGES_REQUESTED** [review]
Great implementation, the new architecture is very well done! Requested changes are just code questions and design challenges mostly.

File: src/remote_gym/remote_environment_management.py
Comment (alexander-zap) [reply src/remote_gym/remote_environment_management.py]: What is your opinion on that? Maybe something like `create_server_for_environment_servicing`?

File: src/remote_gym/remote_environment_management.py
Comment (alexander-zap) [inline src/remote_gym/remote_environment_management.py]: I think we might want to update this part of the documentation as well.

File: src/remote_gym/remote_environment_management.py
Comment (Luke100000) [reply src/remote_gym/remote_environment_management.py]: Might want to use it, specifically environments which require unique resources like ports. Theoretically I could pass it like render_mode and not merge it directly into args, but after all it IS an arg so...

File: src/remote_gym/remote_environment_management.py
Comment (Luke100000) [reply src/remote_gym/remote_environment_management.py]: Just in case I will enforce relative paths.

File: src/remote_gym/remote_environment_management.py
Comment (Luke100000) [reply src/remote_gym/remote_environment_management.py:291]: yes, cleaning up again in destroy_environment

File: src/remote_gym/remote_environment_management.py
Comment (Luke100000) [reply src/remote_gym/remote_environment_management.py]: Theoretically one could join, leave, join, leave and thus add more and more callbacks. We don't do that, but its possible I believe. Its a small memory leak basically.

File: src/remote_gym/remote_environment_management.py
Comment (Luke100000) [reply src/remote_gym/remote_environment_management.py]: `create_remote_environment_server` (server for servicing is equivalent to just server)

File: src/remote_gym/remote_environment.py
Comment (alexander-zap) [inline src/remote_gym/remote_environment.py]: Maybe we can call it something more understandable like `env_init_kwargs` or `entrypoint_kwargs`.

File: src/remote_gym/remote_environment.py
Comment (alexander-zap) [inline src/remote_gym/remote_environment.py:100]: 1. Can we rename this to `additional_remote_args`?
2. Can we create an example for passing `additional_remote_args` in "start_environment_interaction.py"?

File: src/remote_gym/remote_environment.py
Comment (alexander-zap) [inline src/remote_gym/remote_environment.py]:
```suggestion
            remote_args: Additional arguments sent to the remote environment creation servicer. Overwrites "default" arguments which are already configured when initializing the servicer (in `remote_environment_management.create_remote_environment_server`).
```

File: src/remote_gym/remote_environment.py
Comment (Luke100000) [reply src/remote_gym/remote_environment.py:100]: I believe additional is redundant and slightly misleading (its not additional, its overriding technically). Will add an example.

File: src/remote_gym/remote_environment_management.py
Comment (alexander-zap) [reply src/remote_gym/remote_environment_management.py]: Irrelevant now, since you now only set the callback on `create_world`.

File: src/remote_gym/remote_environment_management.py
Comment (alexander-zap) [inline src/remote_gym/remote_environment_management.py]: This doesn't match the example entrypoint. There we don't accept `env_id`.
Is `env_id` non-optional now or does it belong to `kwargs`?

File: src/remote_gym/remote_environment_management.py
Comment (alexander-zap) [inline src/remote_gym/remote_environment_management.py]: I think this documentation is better placed in `RemoteArgs` class definition.

File: exploration/start_environment_interaction.py
Comment (alexander-zap) [inline exploration/start_environment_interaction.py:75]: How can you define remote_args here without having a non-optional "entrypoint" key?

Comment (alexander-zap) [reply exploration/start_environment_interaction.py:75]: I have not tested it, but does it work?

File: src/remote_gym/remote_environment_management.py
Comment (Luke100000) [reply src/remote_gym/remote_environment_management.py]: In the example all arguments not explicitly listed end up in kwargs. Since the example does not make use of env_id I didn't mentioned it.

File: src/remote_gym/remote_environment_management.py
Comment (alexander-zap) [reply src/remote_gym/remote_environment_management.py]: Understood. Maybe then we need to introduce message classes and have different message types.
This way we can still communicate in the same way, but it looks less hacky by assuming that `None` represents something special in certain situations.

File: exploration/start_environment_interaction.py
Comment (Luke100000) [reply exploration/start_environment_interaction.py:75]: Its optional, the server (start_remote_environment) defines it.

Comment (alexander-zap) [reply exploration/start_environment_interaction.py:75]: According to `RemoteArgs` class it's not. Then you can add a note there that it's optional and mention what will be the default (as defined by start_remote_environment).

File: src/remote_gym/remote_environment_management.py
Comment (Luke100000) [reply src/remote_gym/remote_environment_management.py]: What part? The constructor nor the args in this doc is related to RemoteArgs

File: exploration/start_environment_interaction.py
Comment (Luke100000) [reply exploration/start_environment_interaction.py:75]: clarified it

File: src/remote_gym/remote_environment_management.py
Comment (Luke100000) [reply src/remote_gym/remote_environment_management.py]: None does not represent anything special, it is really just the absence of the specs. Since specs are just directories I can pass empty directories instead of that's tidier, but in any case, those values are always ignored.

File: src/remote_gym/remote_environment_management.py
Comment (alexander-zap) [reply src/remote_gym/remote_environment_management.py]: The details about the entrypoint (that it needs to include the `create_environment`).

File: src/remote_gym/remote_environment_management.py
Comment (alexander-zap) [reply src/remote_gym/remote_environment_management.py]: Where are you enforcing relative paths? Haven't found it yet.

Comment (alexander-zap) [reply src/remote_gym/remote_environment_management.py]: Nevermind, found it.

File: src/remote_gym/remote_environment_management.py
Comment (alexander-zap) [inline src/remote_gym/remote_environment_management.py]:
```suggestion
            raise ValueError(f"Unexpected message in remote-gym communication between servicer and environment execution loop: Expected message with content key {key}, but got following message: {msg}")
```

**Review by alexander-zap — APPROVED** [review]
LGTM, awesome merge request! Thanks for all the incorporated feedback :)

File: src/remote_gym/remote_environment_management.py
Comment (alexander-zap) [inline src/remote_gym/remote_environment_management.py:341]: Why do you set this in step?

File: src/remote_gym/remote_environment_management.py
Comment (alexander-zap) [inline src/remote_gym/remote_environment_management.py:379]: How can we even get here, if we removed it in line 352?

File: src/remote_gym/remote_environment_management.py
Comment (alexander-zap) [reply src/remote_gym/remote_environment_management.py:341]: Seems like it belongs in __init__.

Comment (alexander-zap) [inline src/remote_gym/remote_environment_management.py]: Maybe we can rename it to something like `users_in_environments_to_be_destroyed`?

File: src/remote_gym/remote_environment_management.py
Comment (alexander-zap) [inline src/remote_gym/remote_environment_management.py]:
```suggestion
    def get_message_from_out_queue(self, key: str):
```

File: src/remote_gym/remote_environment_management.py
Comment (alexander-zap) [inline src/remote_gym/remote_environment_management.py:241]: Does it make sense to do the same messaging concept here as we did for the out_queue? So having something like `get_message_from_in_queue`.

Then we could also explicitly check for closing commands instead of checking if `reset is None`.

Comment (Luke100000) [reply src/remote_gym/remote_environment_management.py:241]: Personally I would stay at None, an EOL symbol effectively

Comment (alexander-zap) [reply src/remote_gym/remote_environment_management.py:241]: Sure, let's stay at None. But do you think a named messaging concept like for out_queue makes sense here as well?

Comment (alexander-zap) [reply src/remote_gym/remote_environment_management.py:241]: Will be added in a patch.

**Pull request merged**
